streamcopy/copy.go

```diff
@@ -0,0 +1,57 @@
+package streamcopy
+
+import (
+	"bufio"
+	"io"
+	"time"
+
+	"github.com/containers/image/types"
+)
+
+const megaByte = float64(1024 * 1024)
+const readerSize = 32768
+
+// WithProgress implements io.Copy with a buffered reader, then measures
+// reports to the offsetFunc with the count processed at the interval.
+// If io.EOF is not returned then the error is returned. Otherwise, it is nil.
+// WithProgress closes the writer after finishing.
+func WithProgress(writer io.WriteCloser, reader io.Reader, artifact types.BlobInfo, interval time.Duration, signal chan types.SignalProperties) error {
+	defer writer.Close()
+	rd := bufio.NewReaderSize(reader, readerSize)
+
+	count := uint64(0)
+	buf := make([]byte, readerSize)
+	t := time.Now()
+	for {
+		rn, err := rd.Read(buf)
+		count += uint64(rn)
+
+		if err == io.EOF {
+			if rn > 0 {
+				goto write
+			} else {
+				return nil
+			}
+		}
```

Comment:
No, I am not dogmatic about `goto`, though it does seem to me that a `goto`-less equivalent would be a bit shorter. The objection was really about skipping the final update with 100% progress.

Though, with the point about `io.Reader`  returning `rn >0 && err != nil`, I expect this to be restructured so much that nothing about these particular lines will remain applicable.